File: fe/src/lib/migrateChartTypes.ts
import type { ChartViewConfig } from "@/lib/chartViewConfig";
import { pickChartNativeBodyUi } from "@/lib/chartNativeBodyUi";
import { remapLegacySqlFieldsInChartConfig } from "@/lib/legacySqlFieldAliases";
import { migrateTimelineChartConfig } from "@/lib/migrateTimelineChartType";

type StyleMigration = {
  match: (cfg: ChartViewConfig) => boolean;
  nextType: string;
  nextVariant?: string;
};

const STYLE_VARIANT_MIGRATIONS: StyleMigration[] = [
  { match: (c) => c.chartType === "bar" && c.styleVariant === "stacked", nextType: "bar-stack" },
  { match: (c) => c.chartType === "bar" && c.styleVariant === "grouped", nextType: "bar-group" },
  { match: (c) => c.chartType === "bar" && c.styleVariant === "horizontal", nextType: "bar-horizontal" },
  { match: (c) => c.chartType === "pie" && c.styleVariant === "donut", nextType: "pie-donut" },
  { match: (c) => c.chartType === "pie" && c.styleVariant === "rose", nextType: "pie-rose" },
  { match: (c) => c.chartType === "line" && c.styleVariant === "area", nextType: "area" },
  { match: (c) => c.chartType === "line" && c.styleVariant === "stacked", nextType: "area-stack" },
];

const TYPE_MIGRATIONS: Record<string, string> = {
  table: "table-info",
  combo: "chart-mix",
  heatmap: "t-heatmap",
  wordCloud: "word-cloud",
};

/** 将存量 chartType / styleVariant 迁移为 DE 独立 type */
export function migrateChartViewConfig<T extends ChartViewConfig>(config: T): T {
  let next = migrateTimelineChartConfig(config) as T;

  const directTarget = TYPE_MIGRATIONS[next.chartType];
  if (directTarget) {
    next = { ...next, chartType: directTarget as T["chartType"], styleVariant: "default" };
  }

  for (const rule of STYLE_VARIANT_MIGRATIONS) {
    if (!rule.match(next)) continue;
    next = {
      ...next,
      chartType: rule.nextType as T["chartType"],
      styleVariant: rule.nextVariant ?? "default",
    };
    break;
  }

  next = remapLegacySqlFieldsInChartConfig({
    ...next,
    mode: "dataset",
    bindingId: undefined,
    sql: undefined,
    schema: undefined,
    table: undefined,
    index: undefined,
    nativeBody: pickChartNativeBodyUi(next.nativeBody),
  });

  return next;
}

/** 递归迁移 layoutJson 内 widget chartConfig */
export function migrateLayoutChartTypes(layoutJson: unknown): unknown {
  if (!layoutJson || typeof layoutJson !== "object") return layoutJson;
  if (Array.isArray(layoutJson)) {
    return layoutJson.map(migrateLayoutChartTypes);
  }
  const obj = layoutJson as Record<string, unknown>;
  const next: Record<string, unknown> = {};
  for (const [key, value] of Object.entries(obj)) {
    if (key === "chartConfig" && value && typeof value === "object") {
      next[key] = migrateChartViewConfig(value as ChartViewConfig);
      continue;
    }
    next[key] = migrateLayoutChartTypes(value);
  }
  return next;
}

File: backend/app/viz/migrate_chart_types.py
"""存量 chartType / styleVariant → DE 独立 type（与 fe/src/lib/migrateChartTypes.ts 对齐）。"""

from __future__ import annotations

from typing import Any

from app.dashboard.templates.legacy_sql_field_aliases import remap_legacy_sql_fields_in_chart_config

_TYPE_MIGRATIONS: dict[str, str] = {
    "table": "table-info",
    "combo": "chart-mix",
    "heatmap": "t-heatmap",
    "wordCloud": "word-cloud",
}

_STYLE_VARIANT_RULES: tuple[tuple[str, str, str], ...] = (
    ("bar", "stacked", "bar-stack"),
    ("bar", "grouped", "bar-group"),
    ("bar", "horizontal", "bar-horizontal"),
    ("pie", "donut", "pie-donut"),
    ("pie", "rose", "pie-rose"),
    ("line", "area", "area"),
    ("line", "stacked", "area-stack"),
)


def migrate_chart_config(data: Any) -> Any:
    if not isinstance(data, dict):
        return data
    out = dict(data)
    chart_type = out.get("chartType")
    if not isinstance(chart_type, str):
        return out

    direct = _TYPE_MIGRATIONS.get(chart_type)
    if direct:
        out["chartType"] = direct
        out["styleVariant"] = "default"
        chart_type = direct

    style_variant = out.get("styleVariant") or "default"
    if isinstance(style_variant, str):
        for src_type, src_variant, next_type in _STYLE_VARIANT_RULES:
            if chart_type == src_type and style_variant == src_variant:
                out["chartType"] = next_type
                out["styleVariant"] = "default"
                break

    return remap_legacy_sql_fields_in_chart_config(out)


def migrate_layout_chart_configs(layout: Any) -> Any:
    if not layout or not isinstance(layout, dict):
        return layout
    widgets = layout.get("widgets")
    if not isinstance(widgets, list):
        return layout
    next_layout = dict(layout)
    next_widgets: list[Any] = []
    for widget in widgets:
        if not isinstance(widget, dict):
            next_widgets.append(widget)
            continue
        w = dict(widget)
        if w.get("type") == "chart" and isinstance(w.get("chartConfig"), dict):
            w["chartConfig"] = migrate_chart_config(w["chartConfig"])
        next_widgets.append(w)
    next_layout["widgets"] = next_widgets
    return next_layout
